### ISSUE TITLE
PSCX conflicts with multiple commands in version 5.1

### ISSUE THREAD

**Issue by nkrepo** [Pscx/Pscx#23]
PSCX conflicts with multiple commands in version 5.1 build 15063  revision 138

install-package PSCX

```
The package(s) come(s) from a package source that is not marked as trusted.
Are you sure you want to install software from 'PSGallery'?
[Y] Yes  [A] Yes to All  [N] No  [L] No to All  [S] Suspend  [?] Help (default is "N"): y
install-package : The following commands are already available on this system:'gcb,Expand-Archive,Format-Hex,Get-Hash,help,Invoke-NullCoalescing,prompt,Get-Clipboard,Get-Help,Set-Clipboard'. This
module 'Pscx' may override the existing commands. If you still want to install this module 'Pscx', use -AllowClobber parameter.

```

**Comment by codecontemplator** [comment]
+1

**Comment by deuxbits** [comment]
What's going on with this? Is PSCX module compatible with Powershell 5.1 (5.1.14409.1005)? How do you workaround this problem? Is there anything I can do to help?

**Comment by rkeithhill** [comment]
PSCX is compatible with PS v3 through v5.1.  The above is just a **warning**.  Keep in mind that PSCX had all of these commands **years** before they were added to PowerShell.  

Since our commands sometimes have different parameters (or behave differently) removing them is tricky because we can break folks who have written scripts to use the PSCX versions of these commands.  

Anyway the workaround is to module name qualify the command you want to run e.g.:
```posh
Microsoft.PowerShell.Compression\Expand-Archive
```
Or you could use a module prefix when importing the module e.g.
```posh
Import-Module Pscx -Prefix Pscx
Expand-PscxArchive
```

**Comment by davidski** [comment]
Is there a plan to deprecate and eventually phase out the commands that are now included with the core PoSH runtime?

**Comment by nkrepo** [comment]
Maybe the best solution is to use the PSCX prefix by default to avoid clashes with current and feature Powershell releases.  For compatibility reasons, a version without the PSCX prefix may be also released.

**Comment by rkeithhill** [comment]
@davidski The current plan is to let PSCX 3.2.x ride as-is.  Well, I have at least one more update for 3.2.x I need to get out.  After that, we want to subdivide PSCX into smaller modules and make as many as possible run cross-platform.  Just been waiting on the .NET Standard / Core 2.0 tooling in VS 2017, as well as the PS Core ref assemblies, to get to a decent state.  I spent a lot of time on the ASP.NET Core alpha wave porting a web site and got burned by having to fix stuff after just about every new drop.  And then .NET Core 2.0 was announced and a good chunk of that stuff I was changing to work with 1.x didn't need to be changed anymore.  Decided that I wasn't going to spend any more time on .NET Core until 2.0 and good tooling were released.

**Comment by veleek** [comment]
@rkeithhill Do you have any thoughts around breaking up PSCX?  Would the plan be to make a new project for each module under the PSCX GitHub account?  Would it just be a single module for each of the PSCX "sub-modules" or something similar?

**Comment by franklinyu** [comment]
> The current plan is to let PSCX 3.2.x ride as-is.

Does this imply that we do plan to remove them in PSCX 4.x? (which means that PSCX 4.x will not be compatible with PowerShell 4.0 or below.)

**Comment by rkeithhill** [comment]
Remove duplicates yes but that doesn't necessarily mean it will not work on PS 3/4.  However, I'm considering creating a single, portable binary that would load into PowerShell Core.  It will load into Window PowerShell as well except only on systems with .NET 4.7.1 installed.

**Comment by franklinyu** [comment]
.NET 4.7.1? Why? Is it .NET Standard 2.0?

**Comment by rkeithhill** [comment]
Yup. If you want a single, portable binary you compile it as a netstandard2.0 lib.  Otherwise, you have to multi-target to generate both a netstandard2.0 (or netcoreapp2.0 lib) and a net461 lib.  And then figure out which to load at runtime.  All of which is doable but the notion of having a single, portable is more appealing.  Whether that works out in practice remains to be seen.

**Comment by franklinyu** [comment]
I'm not familiar with .NET Framework 4.7.1, but does it drop support for .NET Standard 1.x series? I thought .NET Framework 4.7.1 supports .NET Standard 2 *in addition to* .NET 1.x series. Dropping all the current .NET Standards doesn't sound very Microsoft…

**Comment by rkeithhill** [comment]
No it doesn't drop support.  The thing that .NET 4.7.1 does provide is the ability handle type forwarding for netstandard2.0 libraries to .NET 4.7.1 without requiring recompilation of .NET exe (or tweaks to the app.config to add binding redirects and add probing paths).

**Comment by franklinyu** [comment]
Sorry, but I totally got confused although I know that this is probably basic in .NET area: if the new PSCX is written (and compiled) in .NET Standard 1.2 (to support up to Windows 8.1), will it run on a Windows 10 machine with PowerShell running in .NET Framework 4.7.1? If no, why? Shouldn't PowerShell with .NET Framework 4.7.1 accepts module either in .NET Standard 2.0 or in 1.x?

**Comment by ganziqim** [comment]
+1

**Comment by NielsGrove** [comment]
As suggested by @u0m3 in issue #14 a quick fix can be to import the module with a prefix. That require the module being saved locally first:
`Save-Module -Name 'Pscx' -LiteralPath 'C:\temp\' -Repository PSGallery -RequiredVersion 3.3.1 -Force
Import-Module -Name 'C:\temp\Pscx' -Prefix Pscx -Scope Global -PassThru

Get-Module -Name 'Pscx'
Get-Command -Module 'Pscx'
Remove-Item -LiteralPath 'C:\temp\Pscx' -Force`
Unfortunately it looks like the administrator context is lost before the call of Remove-Item. A new PowerShell session as administrator can remove the folder with its files.

**Comment by Rugby-Ball** [comment]
Can you use -prefix with Install-Module cmdlet? I know its for Import-Module but want to use the nuget repository to install as this will be an on Start-Up script for setting up new servers.
So for example below wont work but, is there any other method?

Install-Module Pscx -Scope CurrentUser -Prefix Pscx -Force

**Comment by subash544772** [comment]
any workaround for this issue yet ?

**Comment by MateuszStepniak** [comment]
@subash544772 for me installing the module with a prefix, as described by @NielsGrove, worked fine.

**Comment by macmodrov** [comment]
Hey, I'm learning PowerShell and wanted to use a recipe from internet which uses Get-Hash. The recipe advises to install and import Pscx.

I cannot do that for reasons cited in this issue. NuGet explicitly says that Pscx conflicts with Get-Hash that is already present (asking other cmdlets).

The thing is... Trying to call Get-Hash gives me an unknown command error and calling Find-Command Get-Hash tells me the command can be found in Pscx.

What the what?

**Comment by MateuszStepniak** [comment]
@macmodrov, please consider installing Pscx with a prefix, as @NielsGrove described. That's a known workaround.

**Comment by macmodrov** [comment]
I probably will. I only hoped someone wild also explain to me the catch 44 situation here.

To have Get-Hash I need to install pscx, but I can't install pscx because I have Get-Hah.

**Comment by FranklinYu** [comment]
@macmodrov That’s weird. What version of PowerShell are you using? I’m using `5.1.17763.134` and `Get-Hash` isn’t in the list of conflicting CmdLets.

**Comment by MateuszStepniak** [comment]
I agree it's weird. I have PowerShell 5.1.17134.590. I didn't have PSCX installed at my home environment and before trying to install it right now, Get-Command was only showing the presence of Get-FileHash command, no Get-Hash one. Still, PSCX installation failed listing Get-Hash as one of the conflicting commands.

**Comment by rkeithhill** [comment]
The error message should indicate that you can use `-AllowClobber` to allow the installation to continue.

**Comment by apaladi** [comment]
> As suggested by @u0m3 in issue #14 a quick fix can be to import the module with a prefix. That require the module being saved locally first:
> `Save-Module -Name 'Pscx' -LiteralPath 'C:\temp' -Repository PSGallery -RequiredVersion 3.3.1 -Force
> Import-Module -Name 'C:\temp\Pscx' -Prefix Pscx -Scope Global -PassThru

After I do that, I get errors when calling cmdlets that use the original name internally:
`Edit-PscxProfile`
Edit-File : The term 'Edit-File' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the spelling of the name, or if a path was included, verify that the path is correct and try again.
At C:\Temp\Pscx\3.3.2\Modules\Utility\Pscx.Utility.psm1:424 char:5

**Comment by fgimian** [comment]
Having some dramas with this approach on my end 😢 

```
PS C:\Users\fots> cd ~
PS C:\Users\fots> Import-Module Pscx -Prefix Pscx
PS C:\Users\fots> cd ~
cd : The term 'Pscx\Set-LocationEx' is not recognized as the name of a cmdlet, function, script file, or operable program. Check the
spelling of the name, or if a path was included, verify that the path is correct and try again.
At line:1 char:1
+ cd ~
+ ~~
    + CategoryInfo          : ObjectNotFound: (Pscx\Set-LocationEx:String) [], CommandNotFoundException
    + FullyQualifiedErrorId : CommandNotFoundException
```

Any ideas?
Fotis

**Comment by fgimian** [comment]
I ended up using each of the functions I needed by simply prefixing them with `Pscx\<function`

e.g.

```powershell
Pscx\Read-Archive -Path $Path -IncludeDirectories
```

The archive expanding facilities in this library are a lifesaver, thank you!!! 😄 
Fotis

**Comment by macmodrov** [comment]
```
PS C:\> Install-Module Pscx -Scope CurrentUser -Prefix Pscx
Install-Module : A parameter cannot be found that matches parameter name 'Prefix'.
At line:1 char:40
+ Install-Module Pscx -Scope CurrentUser -Prefix Pscx
+                                        ~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Install-Module], ParameterBindingException
    + FullyQualifiedErrorId : NamedParameterNotFound,Install-Module
```
PSVersion                      5.1.17763.592

**Comment by FranklinYu** [comment]
@macmodrov Pay attention to difference between `Install-Module` and `Import-Module`.

**Comment by rkeithhill** [comment]
v4 will address most if not all of these.  There is a beta1 I published that addresses most except for help, Get-Help and prompt.  I think I'll remove the prompt functionality (there are way better pkgs now like oh-my-posh).

**Comment by ratijas** [comment]
I don't know what v4 has changed, but I can't seem to use the basic `cd` command alias with Pscx installed.

> Edit: Appears to be due to all the aliases that are imported

So, what should users do about it? Can the module be imported without aliases at all? Because I tried `Import-Module Pscx -Prefix Pscx` and it didn't help.

---

Update: I managed to import only the required function in isolation from the module:

```powershell
Import-Module Pscx -Prefix Pscx -NoClobber -Function Invoke-BatchFile
```
